File: sdk/lx-trading-cpp/src/arbitrage/cross_chain.cpp
// LX Trading SDK - Cross-Chain Router Implementation

#include <lx/trading/arbitrage/cross_chain.hpp>
#include <algorithm>

namespace lx::trading::arbitrage {

CrossChainRouter::CrossChainRouter(CrossChainConfig config)
    : config_(std::move(config)) {}

void CrossChainRouter::set_warp_client(std::shared_ptr<WarpClient> client) {
    std::lock_guard<std::mutex> lock(warp_mutex_);
    warp_client_ = std::move(client);
}

void CrossChainRouter::set_teleport_client(std::shared_ptr<TeleportClient> client) {
    std::lock_guard<std::mutex> lock(teleport_mutex_);
    teleport_client_ = std::move(client);
}

std::shared_ptr<WarpClient> CrossChainRouter::warp() const {
    std::lock_guard<std::mutex> lock(warp_mutex_);
    return warp_client_;
}

std::shared_ptr<TeleportClient> CrossChainRouter::teleport() const {
    std::lock_guard<std::mutex> lock(teleport_mutex_);
    return teleport_client_;
}

CrossChainTransport CrossChainRouter::determine_transport(
    const std::string& source_chain,
    const std::string& dest_chain) const {

    auto src_it = config_.chains.find(source_chain);
    auto dst_it = config_.chains.find(dest_chain);

    // Same chain = direct
    if (source_chain == dest_chain) {
        return CrossChainTransport::Direct;
    }

    // CEX = API
    if (src_it != config_.chains.end() && src_it->second.chain_type == ChainType::Cex) {
        return CrossChainTransport::CexApi;
    }
    if (dst_it != config_.chains.end() && dst_it->second.chain_type == ChainType::Cex) {
        return CrossChainTransport::CexApi;
    }

    // Both Lux subnets = Warp (fastest)
    if (src_it != config_.chains.end() && dst_it != config_.chains.end()) {
        const auto& src = src_it->second;
        const auto& dst = dst_it->second;

        if (src.chain_type == ChainType::LuxSubnet &&
            dst.chain_type == ChainType::LuxSubnet &&
            src.warp_supported && dst.warp_supported &&
            config_.warp_enabled) {
            return CrossChainTransport::Warp;
        }
    }

    // Both EVM or mixed = Teleport
    if (src_it != config_.chains.end() && dst_it != config_.chains.end()) {
        const auto& src = src_it->second;
        const auto& dst = dst_it->second;

        if (src.teleport_supported && dst.teleport_supported && config_.teleport_enabled) {
            return CrossChainTransport::Teleport;
        }
    }

    // No viable transport - return Direct as fallback
    return CrossChainTransport::Direct;
}

int64_t CrossChainRouter::estimate_latency(
    const std::string& source_chain,
    const std::string& dest_chain) const {

    CrossChainTransport transport = determine_transport(source_chain, dest_chain);

    switch (transport) {
        case CrossChainTransport::Direct:
            return 0;
        case CrossChainTransport::Warp:
            return 500;  // Sub-second
        case CrossChainTransport::CexApi:
            return 100;  // API call
        case CrossChainTransport::Teleport: {
            auto src_it = config_.chains.find(source_chain);
            if (src_it != config_.chains.end()) {
                return static_cast<int64_t>(src_it->second.finality_ms) + 10000;  // Finality + processing
            }
            return 3600000;  // 1 hour fallback
        }
    }
    return 3600000;
}

Decimal CrossChainRouter::estimate_cost(
    const std::string& source_chain,
    const std::string& dest_chain,
    const std::string& token,
    Decimal amount) const {

    CrossChainTransport transport = determine_transport(source_chain, dest_chain);

    switch (transport) {
        case CrossChainTransport::Direct:
            return Decimal::zero();
        case CrossChainTransport::Warp:
            return Decimal::from_double(0.001);  // Nearly free
        case CrossChainTransport::CexApi:
            return Decimal::zero();  // No bridge cost
        case CrossChainTransport::Teleport: {
            std::lock_guard<std::mutex> lock(teleport_mutex_);
            if (teleport_client_) {
                try {
                    return teleport_client_->estimate_bridge_fee(dest_chain, token, amount);
                } catch (...) {
                    return Decimal::from_double(1.0);  // Fallback estimate
                }
            }
            return Decimal::from_double(1.0);  // Estimate $1
        }
    }
    return Decimal::zero();
}

std::string CrossChainRouter::venue_to_chain(const std::string& venue) const {
    for (const auto& [chain_id, info] : config_.chains) {
        for (const auto& v : info.venues) {
            if (v == venue) {
                return chain_id;
            }
        }
    }
    return venue;  // Fallback to venue name
}

EnhancedOpportunity CrossChainRouter::enhance_opportunity(const UnifiedOpportunity& opp) const {
    std::string buy_chain = venue_to_chain(opp.buy_venue);
    std::string sell_chain = venue_to_chain(opp.sell_venue);

    CrossChainTransport transport = determine_transport(buy_chain, sell_chain);
    int64_t estimated_latency = estimate_latency(buy_chain, sell_chain);
    Decimal bridge_cost = estimate_cost(buy_chain, sell_chain, opp.symbol, opp.max_size);

    return EnhancedOpportunity{
        .base = opp,
        .transport = transport,
        .estimated_latency = estimated_latency,
        .bridge_cost = bridge_cost,
        .adjusted_net_profit = opp.net_profit - bridge_cost
    };
}

}  // namespace lx::trading::arbitrage

File: sdk/lx-trading-core/src/arbitrage/cross_chain.rs
//! Cross-Chain Arbitrage Transports.
//!
//! 1. WARP (Lux Native)
//!    - Only works WITHIN Lux ecosystem (between subnets)
//!    - Sub-second message delivery (<500ms)
//!    - Use for: LX DEX <-> LX AMM <-> Other Lux subnets
//!    - Cannot reach external chains
//!
//! 2. TELEPORT (EVM Bridge)
//!    - Works with ANY EVM-compatible chain
//!    - Lux <-> Ethereum, BSC, Arbitrum, Polygon, etc.
//!    - ~30 second finality (depends on source chain)
//!    - Uses validator attestations
//!
//! 3. CEX API
//!    - No bridging needed - just API calls
//!    - Sub-second execution
//!    - Settlement via withdraw/deposit (slow but doesn't block arb)
//!
//! 4. FOR OMNICHAIN ARBITRAGE:
//!    - Lux internal: Warp (instant)
//!    - External EVM: Teleport (~30s)
//!    - CEX: Direct API (instant trade, later settle)

use crate::arbitrage::types::*;
use async_trait::async_trait;
use rust_decimal::Decimal;
use std::sync::Arc;
use tokio::sync::RwLock;

/// Warp client interface for Lux-native messaging.
#[async_trait]
pub trait WarpClient: Send + Sync {
    /// Send a Warp message to another Lux subnet.
    async fn send_message(&self, dest_subnet: &str, payload: &[u8]) -> Result<String, String>;

    /// Receive a Warp message.
    async fn receive_message(&self, message_id: &str) -> Result<Vec<u8>, String>;

    /// Get this subnet's ID.
    fn get_blockchain_id(&self) -> String;
}

/// Teleport client interface for EVM bridging.
#[async_trait]
pub trait TeleportClient: Send + Sync {
    /// Bridge assets to another EVM chain.
    async fn bridge(
        &self,
        dest_chain: &str,
        token: &str,
        amount: Decimal,
    ) -> Result<String, String>;

    /// Get bridge transaction status.
    async fn get_bridge_status(&self, tx_id: &str) -> Result<BridgeStatus, String>;

    /// Estimate bridge fee.
    async fn estimate_bridge_fee(
        &self,
        dest_chain: &str,
        token: &str,
        amount: Decimal,
    ) -> Result<Decimal, String>;
}

/// Cross-chain router for determining optimal transport.
pub struct CrossChainRouter {
    config: CrossChainConfig,
    warp: Arc<RwLock<Option<Arc<dyn WarpClient>>>>,
    teleport: Arc<RwLock<Option<Arc<dyn TeleportClient>>>>,
}

impl CrossChainRouter {
    /// Create a new cross-chain router.
    pub fn new(config: CrossChainConfig) -> Self {
        Self {
            config,
            warp: Arc::new(RwLock::new(None)),
            teleport: Arc::new(RwLock::new(None)),
        }
    }

    /// Set the Warp client.
    pub async fn set_warp_client(&self, client: Arc<dyn WarpClient>) {
        let mut warp = self.warp.write().await;
        *warp = Some(client);
    }

    /// Set the Teleport client.
    pub async fn set_teleport_client(&self, client: Arc<dyn TeleportClient>) {
        let mut teleport = self.teleport.write().await;
        *teleport = Some(client);
    }

    /// Get the Warp client.
    pub async fn warp(&self) -> Option<Arc<dyn WarpClient>> {
        let warp = self.warp.read().await;
        warp.clone()
    }

    /// Get the Teleport client.
    pub async fn teleport(&self) -> Option<Arc<dyn TeleportClient>> {
        let teleport = self.teleport.read().await;
        teleport.clone()
    }

    /// Determine the best transport between two chains.
    pub fn determine_transport(&self, source_chain: &str, dest_chain: &str) -> CrossChainTransport {
        let src = self.config.chains.get(source_chain);
        let dst = self.config.chains.get(dest_chain);

        // Same chain = direct
        if source_chain == dest_chain {
            return CrossChainTransport::Direct;
        }

        // CEX = API
        if let Some(s) = src {
            if s.chain_type == ChainType::Cex {
                return CrossChainTransport::CexApi;
            }
        }
        if let Some(d) = dst {
            if d.chain_type == ChainType::Cex {
                return CrossChainTransport::CexApi;
            }
        }

        // Both Lux subnets = Warp (fastest)
        if let (Some(s), Some(d)) = (src, dst) {
            if s.chain_type == ChainType::LuxSubnet
                && d.chain_type == ChainType::LuxSubnet
                && s.warp_supported
                && d.warp_supported
                && self.config.warp_enabled
            {
                return CrossChainTransport::Warp;
            }
        }

        // Both EVM or mixed = Teleport
        if let (Some(s), Some(d)) = (src, dst) {
            if s.teleport_supported && d.teleport_supported && self.config.teleport_enabled {
                return CrossChainTransport::Teleport;
            }
        }

        // No viable transport - return Direct as fallback
        CrossChainTransport::Direct
    }

    /// Estimate latency for cross-chain message (ms).
    pub fn estimate_latency(&self, source_chain: &str, dest_chain: &str) -> i64 {
        let transport = self.determine_transport(source_chain, dest_chain);

        match transport {
            CrossChainTransport::Direct => 0,
            CrossChainTransport::Warp => 500,   // Sub-second
            CrossChainTransport::CexApi => 100, // API call
            CrossChainTransport::Teleport => {
                let src = self.config.chains.get(source_chain);
                src.map(|s| s.finality_ms as i64 + 10000).unwrap_or(3600000)
                // Finality + processing
            }
        }
    }

    /// Estimate cost for cross-chain transfer.
    pub async fn estimate_cost(
        &self,
        source_chain: &str,
        dest_chain: &str,
        token: &str,
        amount: Decimal,
    ) -> Decimal {
        let transport = self.determine_transport(source_chain, dest_chain);

        match transport {
            CrossChainTransport::Direct => Decimal::ZERO,
            CrossChainTransport::Warp => Decimal::new(1, 3), // 0.001 - Nearly free
            CrossChainTransport::CexApi => Decimal::ZERO,    // No bridge cost
            CrossChainTransport::Teleport => {
                let teleport = self.teleport.read().await;
                if let Some(ref client) = *teleport {
                    client
                        .estimate_bridge_fee(dest_chain, token, amount)
                        .await
                        .unwrap_or_else(|_| Decimal::from(1))
                } else {
                    Decimal::from(1) // Estimate $1
                }
            }
        }
    }

    /// Get chain ID from venue name.
    pub fn venue_to_chain(&self, venue: &str) -> String {
        for (chain_id, info) in &self.config.chains {
            if info.venues.contains(&venue.to_string()) {
                return chain_id.clone();
            }
        }
        venue.to_string() // Fallback to venue name
    }

    /// Enhance an opportunity with routing information.
    pub async fn enhance_opportunity(&self, opp: UnifiedOpportunity) -> EnhancedOpportunity {
        let buy_chain = self.venue_to_chain(&opp.buy_venue);
        let sell_chain = self.venue_to_chain(&opp.sell_venue);

        let transport = self.determine_transport(&buy_chain, &sell_chain);
        let estimated_latency = self.estimate_latency(&buy_chain, &sell_chain);
        let bridge_cost = self
            .estimate_cost(&buy_chain, &sell_chain, &opp.symbol, opp.max_size)
            .await;

        EnhancedOpportunity {
            base: opp.clone(),
            transport,
            estimated_latency,
            bridge_cost,
            adjusted_net_profit: opp.net_profit - bridge_cost,
        }
    }

    /// Get current config.
    pub fn config(&self) -> &CrossChainConfig {
        &self.config
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_determine_transport_same_chain() {
        let config = CrossChainConfig::default();
        let router = CrossChainRouter::new(config);

        let transport = router.determine_transport("lux_mainnet", "lux_mainnet");
        assert_eq!(transport, CrossChainTransport::Direct);
    }

    #[test]
    fn test_determine_transport_lux_subnets() {
        let config = CrossChainConfig::default();
        let router = CrossChainRouter::new(config);

        let transport = router.determine_transport("lux_mainnet", "lx_dex_subnet");
        assert_eq!(transport, CrossChainTransport::Warp);
    }

    #[test]
    fn test_determine_transport_cex() {
        let config = CrossChainConfig::default();
        let router = CrossChainRouter::new(config);

        let transport = router.determine_transport("binance", "lux_mainnet");
        assert_eq!(transport, CrossChainTransport::CexApi);
    }

    #[test]
    fn test_determine_transport_evm() {
        let config = CrossChainConfig::default();
        let router = CrossChainRouter::new(config);

        let transport = router.determine_transport("ethereum", "arbitrum");
        assert_eq!(transport, CrossChainTransport::Teleport);
    }

    #[test]
    fn test_estimate_latency() {
        let config = CrossChainConfig::default();
        let router = CrossChainRouter::new(config);

        // Direct = 0
        let latency = router.estimate_latency("lux_mainnet", "lux_mainnet");
        assert_eq!(latency, 0);

        // Warp = 500ms
        let latency = router.estimate_latency("lux_mainnet", "lx_dex_subnet");
        assert_eq!(latency, 500);

        // CEX = 100ms
        let latency = router.estimate_latency("binance", "lux_mainnet");
        assert_eq!(latency, 100);
    }

    #[test]
    fn test_venue_to_chain() {
        let config = CrossChainConfig::default();
        let router = CrossChainRouter::new(config);

        let chain = router.venue_to_chain("lx_dex");
        assert!(chain == "lux_mainnet" || chain == "lx_dex_subnet");

        let chain = router.venue_to_chain("binance");
        assert_eq!(chain, "binance");

        let chain = router.venue_to_chain("uniswap");
        assert!(chain == "ethereum" || chain == "arbitrum");
    }
}
